File: sdk/java/src/main/java/jp/kengos/shojiku/ExternalSigner.java
package jp.kengos.shojiku;

import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.Base64;
import java.util.function.UnaryOperator;

/**
 * A signing provider for a key this process is never given.
 *
 * <p>The second provider, and the shape {@link LocalPem}'s own comment promised: a new class rather
 * than new arguments on {@code sign}, so the call site is unchanged in all seven SDKs.
 *
 * <p>The engine hands out the bytes a signature has to cover; the function signs them wherever the
 * key actually lives — AWS KMS, Google Cloud KMS, an HSM, a smartcard, another service entirely —
 * and hands the signature back:
 *
 * <pre>{@code
 * var provider = ExternalSigner.of(
 *     toBeSigned -> kms.asymmetricSign(keyName, Digest.newBuilder()...).getSignature().toByteArray(),
 *     Algorithm.ECDSA_P256_SHA256,
 *     Path.of("signer.crt"));
 * client.sign(artifact, provider);
 * }</pre>
 *
 * <p>Shojiku ships no cloud client of its own, deliberately: the function is whatever client your
 * application already has, and the SDK stays a wrapper with nothing to keep in step with a vendor's
 * releases.
 *
 * <p><b>What the function receives is the signed ATTRIBUTES, not the document digest.</b> A service
 * that signs a digest must hash these bytes with SHA-256 itself. Signing the document digest
 * instead produces a document that fails verification, so the distinction is not cosmetic.
 *
 * <p>Nothing here is key material — that is the point of this provider — but the function closes
 * over whatever built it, which in practice is a client holding credentials. So {@link #toString}
 * states the certificate's FORM and the algorithm and nothing else, exactly as {@link LocalPem}
 * does.
 */
public final class ExternalSigner implements EngineSigner {

  private final UnaryOperator<byte[]> sign;
  private final Algorithm algorithm;
  private final Path certPath;
  private byte[] certPem;

  private ExternalSigner(
      UnaryOperator<byte[]> sign, Algorithm algorithm, Path certPath, byte[] certPem) {
    if (sign == null) {
      throw new UsageException("ExternalSigner needs a function that signs the bytes it is given");
    }
    if (algorithm == null) {
      throw new UsageException("ExternalSigner needs an algorithm");
    }
    requireCertificate(certPath, certPem);
    this.sign = sign;
    this.algorithm = algorithm;
    this.certPath = certPath;
    this.certPem = certPem;
  }

  /**
   * A provider whose certificate is read from a path.
   *
   * @param sign receives the bytes to sign, returns the raw signature
   * @param algorithm which algorithm the key signs with
   * @param cert path to the signer's certificate, as PEM
   * @return the provider
   */
  public static ExternalSigner of(UnaryOperator<byte[]> sign, Algorithm algorithm, Path cert) {
    return new ExternalSigner(sign, algorithm, cert, null);
  }

  /**
   * A provider whose certificate is already in memory, so it never has to be written to disk.
   *
   * @param sign receives the bytes to sign, returns the raw signature
   * @param algorithm which algorithm the key signs with
   * @param certPem the certificate as bytes
   * @return the provider
   */
  public static ExternalSigner ofPem(
      UnaryOperator<byte[]> sign, Algorithm algorithm, byte[] certPem) {
    return new ExternalSigner(sign, algorithm, null, certPem);
  }

  /**
   * The algorithm this provider's key signs with.
   *
   * @return the algorithm
   */
  public Algorithm algorithm() {
    return algorithm;
  }

  /**
   * The signing certificate, as PEM or DER bytes.
   *
   * @return the certificate material
   */
  public byte[] certificate() {
    if (certPem == null) {
      certPem = Texts.readMaterial(certPath, "certificate_unreadable");
    }
    return certPem;
  }

  /**
   * Signs in two engine calls, with the caller's function in between.
   *
   * <p>Both calls take the same document, certificate and algorithm: the pair is stateless, so the
   * second re-derives what the first prepared. Keeping them inside ONE method is what makes that
   * impossible to get wrong from Java — there is no way to pair a prepare of one document with a
   * complete of another.
   *
   * <p>A prepare that did not succeed is returned as it is: an unreadable certificate or a document
   * the engine refuses is a fact about the inputs, and paying for a signature afterwards would tell
   * the caller nothing new.
   */
  @Override
  public Snapshot signWith(Engine engine, byte[] pdf) {
    byte[] certificate = certificate();
    byte[] wire = algorithm.wire().getBytes(StandardCharsets.UTF_8);
    Snapshot prepared = engine.signPrepare(pdf, certificate, wire);
    if (prepared.status() != 0 || !prepared.success()) {
      return prepared;
    }
    return engine.signComplete(pdf, certificate, wire, signatureFor(prepared));
  }

  /**
   * The bytes the engine wants signed, out of the prepare payload.
   *
   * <p>A method of its own so the refusal is reachable from a test: the real engine always reports
   * {@code toBeSigned}, so a payload without one is a shape only a different library on the other
   * end could produce — and a guard nobody can exercise is a guard nobody knows works.
   *
   * @param json the prepare payload
   * @return the bytes to sign
   */
  static byte[] bytesToSign(String json) {
    // `Json.object` answers an empty map rather than null for anything that is
    // not an object, so the one check below covers every shape at once.
    Object encoded = Json.object(json).get("toBeSigned");
    if (!(encoded instanceof String text)) {
      throw new UsageException("the engine reported no bytes to sign");
    }
    return Base64.getDecoder().decode(text);
  }

  /**
   * Runs the function over the bytes the engine wants signed.
   *
   * <p>The function's own exceptions are deliberately not caught: it is the caller's code talking
   * to the caller's key service, and turning its failures into a failed result would file a
   * caller's outage under "something was wrong with this document".
   */
  private byte[] signatureFor(Snapshot prepared) {
    byte[] signature = sign.apply(bytesToSign(prepared.json()));
    if (signature == null || signature.length == 0) {
      throw new UsageException("the signing function must return a non-empty signature");
    }
    return signature;
  }

  /** Redacted, deliberately — see the class javadoc. */
  @Override
  public String toString() {
    String form = certPath == null ? "[pem bytes]" : certPath.toString();
    return "<ExternalSigner cert=" + form + " algorithm=" + algorithm.wire() + ">";
  }

  /**
   * Explicit, never sniffed — {@link LocalPem}'s rule, half of which Java gets for free.
   *
   * <p>{@link LocalPem} refuses "both forms at once" at run time because its constructor takes
   * both. Here the two FACTORIES make that unrepresentable: a caller reaches {@link #of} or {@link
   * #ofPem} and cannot name both, so the only thing left to check is that one of them carried
   * something.
   */
  private static void requireCertificate(Path path, byte[] pem) {
    if (path == null && pem == null) {
      throw new UsageException(
          "ExternalSigner needs either a certificate path or certificate bytes");
    }
  }
}

File: sdk/php/src/ExternalSigner.php
<?php

declare(strict_types=1);

namespace Shojiku;

use Shojiku\Exception\EngineFailureException;
use Shojiku\Exception\UsageException;

/**
 * A signing provider for a key this process is never given.
 *
 * The second provider, and the shape {@see LocalPem}'s own comment promised: a
 * new class rather than new arguments on `sign`, so the call site is unchanged
 * in all seven SDKs.
 *
 * The engine hands out the bytes a signature has to cover; the callable signs
 * them wherever the key actually lives — AWS KMS, Google Cloud KMS, an HSM, a
 * smartcard, another service entirely — and hands the signature back:
 *
 * ```php
 * $provider = new Shojiku\ExternalSigner(
 *     sign: fn (string $toBeSigned): string => $kms->sign([
 *         'KeyId' => getenv('KEY_ID'),
 *         'Message' => $toBeSigned,
 *         'MessageType' => 'RAW',
 *         'SigningAlgorithm' => 'ECDSA_SHA_256',
 *     ])['Signature'],
 *     cert: 'signer.crt',
 *     algorithm: Shojiku\Algorithm::EcdsaP256Sha256,
 * );
 * $client->sign($artifact, $provider);
 * ```
 *
 * Shojiku ships no cloud client of its own, deliberately: the callable is
 * whatever client your application already has, and the SDK stays a wrapper
 * with nothing to keep in step with a vendor's releases.
 *
 * **What the callable receives is the signed ATTRIBUTES, not the document
 * digest.** A service that signs a digest must hash these bytes with SHA-256
 * itself. Signing the document digest instead produces a document that fails
 * verification, so the distinction is not cosmetic.
 *
 * The signature is the raw output of that operation: PKCS#1 v1.5 bytes for
 * `rsa-pkcs1-sha256`, an ASN.1 DER sequence for `ecdsa-p256-sha256` — which is
 * what both major cloud key services return unchanged.
 *
 * **Redaction.** Nothing here is key material — that is the point of this
 * provider — but a callable closes over whatever built it, which in practice
 * is a client holding credentials. So the callable lives in the same private
 * `\WeakMap` {@see LocalPem} keeps its secrets in, which is invisible to all
 * four of PHP's dumpers, and the constructor's callable is marked
 * `#[\SensitiveParameter]` so a stack trace cannot print it either.
 */
final class ExternalSigner implements SigningProvider
{
    /**
     * The callables, keyed by the provider that owns them. Deliberately not a
     * property: see the class note above.
     *
     * @var \WeakMap<self, callable(string): string>|null
     */
    private static ?\WeakMap $signers = null;

    private readonly ?string $certPath;
    private readonly ?string $certPem;
    public readonly Algorithm $algorithm;

    /**
     * @param callable(string): string $sign receives the bytes to sign,
     *                                       returns the raw signature
     */
    public function __construct(
        #[\SensitiveParameter]
        callable $sign,
        ?string $cert = null,
        ?string $certPem = null,
        Algorithm|string|null $algorithm = null,
    ) {
        self::oneSource($cert, $certPem);
        $this->certPath = $cert;
        $this->certPem = $certPem;
        $this->algorithm = self::wireAlgorithm($algorithm);
        self::store()[$this] = $sign;
    }

    /**
     * The redacted printed form — the class, which FORM the certificate came
     * from, and the algorithm. A configured file path is not secret and is
     * the one thing worth seeing when a provider loaded the wrong material.
     *
     * @return array<string, string>
     */
    public function __debugInfo(): array
    {
        return [
            'cert' => $this->certPath ?? '[pem bytes]',
            'algorithm' => $this->algorithm->value,
        ];
    }

    public function __toString(): string
    {
        $shown = $this->__debugInfo();

        return sprintf('%s cert=%s algorithm=%s', self::class, $shown['cert'], $shown['algorithm']);
    }

    /**
     * Signs in two engine calls, with the caller's callable in between.
     *
     * Both calls take the same document, certificate and algorithm: the pair
     * is stateless, so the second re-derives what the first prepared. Keeping
     * them inside ONE method is what makes that impossible to get wrong from
     * PHP — there is no way to pair a prepare of one document with a complete
     * of another.
     *
     * A prepare that did not succeed is returned as it is: an unreadable
     * certificate or a document the engine refuses is a fact about the
     * inputs, and paying for a signature afterwards would tell the caller
     * nothing new.
     *
     * @return array{Report, string}
     */
    public function signWith(Engine $engine, Workspace $workspace, DocumentArtifact $artifact): array
    {
        $engine->requireExternal();
        $input = $workspace->write('input.pdf', $artifact->bytes());
        $cert = $this->certPath ?? $workspace->write('cert.pem', (string) $this->certPem);
        $algorithm = $this->algorithm->value;

        [$prepared] = $engine->execute(Request::signPrepare($input, $cert, $algorithm), $workspace);
        if (!$prepared->ok) {
            return [$prepared, ''];
        }

        return $engine->execute(
            Request::signComplete(
                $input,
                $cert,
                $algorithm,
                $workspace->write('signature.bin', $this->signatureFor($prepared)),
            ),
            $workspace,
        );
    }

    /**
     * Runs the callable over the bytes the engine wants signed.
     *
     * The callable's own exceptions are deliberately not caught: it is the
     * caller's code talking to the caller's key service, and turning its
     * failures into a failed {@see Result} would file a caller's outage under
     * "something was wrong with this document".
     */
    private function signatureFor(Report $prepared): string
    {
        $payload = $prepared->prepared;
        if ($payload === null || !is_string($payload['toBeSigned'] ?? null)) {
            throw new EngineFailureException('the engine reported no bytes to sign');
        }
        $toBeSigned = base64_decode($payload['toBeSigned'], true);
        if ($toBeSigned === false) {
            throw new EngineFailureException("the engine's bytes-to-sign payload is not base64");
        }

        $signature = (self::store()[$this])($toBeSigned);
        if ($signature === '') {
            throw new UsageException('the signing callable must return a non-empty signature');
        }

        return $signature;
    }

    /**
     * @return \WeakMap<self, callable(string): string>
     */
    private static function store(): \WeakMap
    {
        /** @var \WeakMap<self, callable(string): string> */
        return self::$signers ??= new \WeakMap();
    }

    private static function wireAlgorithm(Algorithm|string|null $algorithm): Algorithm
    {
        if ($algorithm instanceof Algorithm) {
            return $algorithm;
        }
        $named = implode(' or ', array_map(
            static fn (Algorithm $case): string => sprintf('`%s`', $case->value),
            Algorithm::cases(),
        ));
        if ($algorithm === null) {
            throw new UsageException(sprintf('ExternalSigner needs `algorithm:` (%s)', $named));
        }

        // The caller's string is never echoed — it came from configuration
        // this package does not control, and the accepted names are the
        // useful half of the answer.
        return Algorithm::tryFrom($algorithm)
            ?? throw new UsageException(sprintf('`algorithm:` must be one of %s', $named));
    }

    /**
     * Explicit, never sniffed — in BOTH directions, {@see LocalPem}'s rule.
     */
    private static function oneSource(?string $path, ?string $pem): void
    {
        $forms = '`cert:` (a path) or `certPem:` (bytes)';
        if ($path !== null && $pem !== null) {
            throw new UsageException(sprintf('ExternalSigner takes either %s, not both', $forms));
        }
        if ($path === null && $pem === null) {
            throw new UsageException(sprintf('ExternalSigner needs either %s', $forms));
        }
    }
}
